src/Configuration/Types.hs

```diff
@@ -0,0 +1,383 @@
+    , renderCluster
+    ) where
+
+import           Cardano.Prelude hiding (evalState)
+
+import           Control.Monad.Trans.State.Strict (evalState)
+import           Data.Functor.Contravariant (contramap)
+import qualified Data.Text as T
+import           Dhall (GenericInject, GenericInterpret, Inject (..), InputType,
+                        Interpret (..), InterpretOptions (..), Type, auto,
+                        field, genericAutoWith, genericInjectWith, record,
+                        strictText)
+import           GHC.Generics (from, to)
+import           Test.QuickCheck (Arbitrary (..), Gen, arbitraryASCIIChar,
+                                  choose, elements, listOf, listOf1)
+
+-- | Operating system
+data OS
+    = Linux64
+    | Macos64
+    | Win64
+    deriving (Bounded, Enum, Eq, Read, Show)
+
+-- | Cluster
+data Cluster
+    = Mainnet
+    | Staging
+    | Testnet
+    | Demo
+    deriving (Bounded, Enum, Eq, Read, Show)
+
+-- | Convert 'OS' into 'Text'
+renderOS :: OS -> Text
+renderOS Linux64 = "linux64"
+renderOS Macos64 = "macos64"
+renderOS Win64   = "win64"
+
+-- | Convert 'Cluster' into 'Text'
+renderCluster :: Cluster -> Text
+renderCluster Mainnet = "mainnet"
+renderCluster Testnet = "testnet"
+renderCluster Staging = "staging"
+renderCluster Demo    = "demo"
+
+-- | Cluster configuration
+data ClusterConfig = ClusterConfig
+    { ccfgName                   :: !Text
+    , ccfgKeyPrefix              :: !Text
+    , ccfgRelays                 :: !Text
+    , ccfgUpdateServer           :: !Text
+    , ccfgReportServer           :: !Text
+    , ccfgInstallDirectorySuffix :: !Text
+    , ccfgMacPackageSuffix       :: !Text
+    , ccfgWalletPort             :: !Natural
+    } deriving (Eq, Generic, Show)
+
+-- Defining each 'Intrepret' instance.
+instance Interpret ClusterConfig where
+    autoWith opt = interpretAutoWithOption $ opt {fieldModifier = lowerHead . T.drop 4}
+
+instance Inject ClusterConfig where
+    injectWith opt = injectAutoWithOption $ opt {fieldModifier = lowerHead . T.drop 4}
+
+
+instance Arbitrary ClusterConfig where
+    arbitrary = do
+        ccfgName                   <- genSafeText
+        ccfgKeyPrefix              <- genSafeText
+        ccfgRelays                 <- genSafeText
+        ccfgUpdateServer           <- genSafeText
+        ccfgReportServer           <- genSafeText
+        ccfgInstallDirectorySuffix <- genSafeText
+        ccfgMacPackageSuffix       <- genSafeText
+        ccfgWalletPort             <- arbitrary
+
+        pure ClusterConfig {..}
+
+-- | OS configuration
+data OSConfig = OSConfig
+    { osName              :: !Text
+    , osConfigurationYaml :: !Text
+    , osInstallDirectory  :: !Text
+    , osX509ToolPath      :: !Text
+    , osNodeArgs          :: !NodeArgs
+    , osPass              :: !Pass
+    } deriving (Eq, Generic, Show)
+
+instance Interpret OSConfig where
+    autoWith opt = interpretAutoWithOption $ opt {fieldModifier = lowerHead . T.drop 2}
```

Comment:
I really don't like these `T.drop` and transformations. It could be very error-prone and detected only on runtime (yes, tests might catch this, but it would be better if we don't have to do that).
We could simply keep things simple and name them _same as in the configuration file_ and be done with it.
The cost is a couple of letters as a prefix, and I consider that to be acceptable.

Reply:
@ksaric In that case, what should I do with the fields that has exact same name? For example `relays`
https://github.com/input-output-hk/cardano-shell/blob/hiroto/reuse-dhall/src/Configuration/Types.hs#L67
https://github.com/input-output-hk/cardano-shell/blob/hiroto/reuse-dhall/src/Configuration/Types.hs#L282

`name`
https://github.com/input-output-hk/cardano-shell/blob/hiroto/reuse-dhall/src/Configuration/Types.hs#L107
https://github.com/input-output-hk/cardano-shell/blob/hiroto/reuse-dhall/src/Configuration/Types.hs#L65

Reply:
No, I meant modify the configuration files names to reflect the names in Haskell.

Reply:
Ahh okay